lib/Compilation.js

```diff
@@ -96,92 +115,156 @@ function addAllToSet(set, otherSet) {
 }
 
 class Compilation extends Tapable {
+	/**
+	 * Creates an instance of Compilation.
+	 * @param {Compiler} compiler the compiler which created the compilation
+	 */
 	constructor(compiler) {
 		super();
 		this.hooks = {
+			/** @type {SyncHook<Module>} */
 			buildModule: new SyncHook(["module"]),
+			/** @type {SyncHook<Module>} */
 			rebuildModule: new SyncHook(["module"]),
+			/** @type {SyncHook<Module, Error>} */
 			failedModule: new SyncHook(["module", "error"]),
+			/** @type {SyncHook<Module>} */
 			succeedModule: new SyncHook(["module"]),
-
+			/** @type {SyncHook<Module[]>} */
 			finishModules: new SyncHook(["modules"]),
+			/** @type {SyncHook<Module>} */
 			finishRebuildingModule: new SyncHook(["module"]),
-
+			/** @type {SyncHook} */
 			unseal: new SyncHook([]),
+			/** @type {SyncHook} */
 			seal: new SyncHook([]),
 
+			/** @type {SyncBailHook<Module[]>} */
 			optimizeDependenciesBasic: new SyncBailHook(["modules"]),
+			/** @type {SyncBailHook<Module[]>} */
 			optimizeDependencies: new SyncBailHook(["modules"]),
+			/** @type {SyncBailHook<Module[]>} */
 			optimizeDependenciesAdvanced: new SyncBailHook(["modules"]),
+			/** @type {SyncBailHook<Module[]>} */
 			afterOptimizeDependencies: new SyncHook(["modules"]),
 
+			/** @type {SyncHook} */
```

Comment:
`SyncHook<void>` would be better I think

Reply:
I guess we don't prevent or want to prevent the overriding of SyncHook<any> right?